### ISSUE TITLE
Support for Linux/macOS

### ISSUE THREAD

**Issue by gtw123** [gtw123/ShenzhenMod#3]
Make a version of the installer that works on Linux and macOS.

**Comment by gtw123** [comment]
I believe Shenzhen.exe is identical on all three platforms (as it is a managed app). So in the meantime the following workaround may work:

* Obtain a Windows machine or VM.
* Copy the whole Shenzhen I/O folder from your Linux or macOS machine to the Windows machine.
* Run the ShenzhenMod installer in Windows.
* In the installer, choose the folder you copied to the machine. Specifically, choose the folder containing Shenzhen.exe.
* Click Install.
* If it succeeds, copy the folder back to the Linux or macOS machine. (It patches some additional files besides Shenzhen.exe so it's easiest if you just copy the whole folder.)

**Comment by 12345ieee** [comment]
Thank you for your work, linux guy here to annoy you some more.

The installer compiles on linux successfully, but a very recent versions of `mono` (>= 5.0) id needed, because this project uses C# 7 features.

The next problem is that you have hardcoded `\` as path separator everywhere, it took a while to fix that, by using `Path.Combine()`.
I haven't touched the messages, nor the GOG path, but those should be fixed as well.

Then I needed a way to detect we're running on linux, and I thank stackoverflow for the `(int) Environment.OSVersion.Platform == 4`.

After doing all this, the patcher ran successfully, but the game crashed on startup with this trace:
```
CALL STACK:

Method with open type while not compiling gshared
  at Solution.#=q3xeqgWSi8bQpaBp419N7hg== () [0x000a8] in <a908011ea2f54d43a49a4ac28cea06b4>:0 
  at Solution.#=qCUAvLv7aG_rY3UiFJy5aGA== (System.String #=q6fen$cGShXnfisSbRV$t_g==) [0x00020] in <a908011ea2f54d43a49a4ac28cea06b4>:0 
  at #=qupCCsXra7736WjYNlnzX8prlJFUjffxo1X9Q6rQqGhM=.#=qGHyGoN0aJvYMEtNq75NsOw== () [0x00029] in <a908011ea2f54d43a49a4ac28cea06b4>:0 
  at GameLogic.#=qo6qy4wPTipgmw9etdEgdNQ== () [0x0095d] in <a908011ea2f54d43a49a4ac28cea06b4>:0 
  at #=qjp0tCZ3dnVicMf2NbL$kGQ==.#=qaEhZyjiKc5xbsK2eqF2CFA== (System.String[] #=qznhcImXm7KyO7Yli3ePJBQ==) [0x0003e] in <a908011ea2f54d43a49a4ac28cea06b4>:0 
```

I can send you the linux exe via some private channel if that helps you.
Also, I attach the diff, so you can verify I didn't break anything:

```diff
diff --git a/ShenzhenMod/Patches/AddBiggerSandbox.cs b/ShenzhenMod/Patches/AddBiggerSandbox.cs
index 3676a12..64bebc1 100644
--- a/ShenzhenMod/Patches/AddBiggerSandbox.cs
+++ b/ShenzhenMod/Patches/AddBiggerSandbox.cs
@@ -219,7 +219,7 @@ namespace ShenzhenMod.Patches
         {
             using (var stream = System.Reflection.Assembly.GetCallingAssembly().GetManifestResourceStream("ShenzhenMod.Content.messages.en.bigger-prototyping-area.txt"))
             {
-                string path = Path.Combine(m_shenzhenDir, @"Content\messages.en\bigger-prototyping-area.txt");
+                string path = Path.Combine(m_shenzhenDir, "Content", "messages.en", "bigger-prototyping-area.txt");
                 using (var file = File.Create(path))
                 {
                     sm_log.InfoFormat("Writing resource file to \"{0}\"", path);
@@ -227,7 +227,7 @@ namespace ShenzhenMod.Patches
                 }
 
                 // Although we haven't got a Chinese version, we need to have a corresponding file in messages.zh to avoid a crash.
-                string path2 = Path.Combine(m_shenzhenDir, @"Content\messages.zh\bigger-prototyping-area.txt");
+                string path2 = Path.Combine(m_shenzhenDir, "Content", "messages.zh", "bigger-prototyping-area.txt");
                 sm_log.InfoFormat("Copying \"{0}\" to \"{1}\"", path, path2);
                 File.Copy(path, path2, overwrite: true);
             }
@@ -235,4 +235,4 @@ namespace ShenzhenMod.Patches
     }
 }
  
- 
\ No newline at end of file
+ 
diff --git a/ShenzhenMod/Patches/AdjustPlaybackSpeedSlider.cs b/ShenzhenMod/Patches/AdjustPlaybackSpeedSlider.cs
index d4623d6..e839761 100644
--- a/ShenzhenMod/Patches/AdjustPlaybackSpeedSlider.cs
+++ b/ShenzhenMod/Patches/AdjustPlaybackSpeedSlider.cs
@@ -63,7 +63,7 @@ namespace ShenzhenMod.Patches
 
             void MakeNewTexture()
             {
-                string editorPath = Path.Combine(m_shenzhenDir, @"Content\textures\editor");
+                string editorPath = Path.Combine(m_shenzhenDir, "Content", "textures", "editor");
                 string panelTexturePath = Path.Combine(editorPath, "panel_sandbox.png");
                 using (var image = Image.FromFile(panelTexturePath))
                 {
diff --git a/ShenzhenMod/ShenzhenLocator.cs b/ShenzhenMod/ShenzhenLocator.cs
index 7398121..6dc62e0 100644
--- a/ShenzhenMod/ShenzhenLocator.cs
+++ b/ShenzhenMod/ShenzhenLocator.cs
@@ -18,11 +18,17 @@ namespace ShenzhenMod
 
         public static string FindShenzhenSteamDirectory()
         {
-            string steamPath = Registry.GetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\Valve\Steam", "InstallPath", null) as string;
+            string steamPath;
+            if ((int) Environment.OSVersion.Platform == 4) { // linux
+                steamPath = Environment.GetEnvironmentVariable("HOME") + @"/.steam/steam";
+            }
+            else {
+                steamPath = Registry.GetValue(@"HKEY_LOCAL_MACHINE\SOFTWARE\Valve\Steam", "InstallPath", null) as string;
+            }
             sm_log.InfoFormat("Steam install path: \"{0}\"", steamPath);
             if (steamPath != null)
             {
-                var shenzhenDir = Path.Combine(steamPath, @"steamapps\common\SHENZHEN IO");
+                var shenzhenDir = Path.Combine(steamPath, "steamapps", "common", "SHENZHEN IO");
                 if (Directory.Exists(shenzhenDir))
                 {
                     sm_log.InfoFormat("Found SHENZHEN I/O directory: \"{0}\"", shenzhenDir);
@@ -109,4 +115,4 @@ namespace ShenzhenMod
             }
         }
     }
-}
\ No newline at end of file
+}
```

**Comment by gtw123** [comment]
I tried it on my mac and it crashes too (although I don't know if it's the same place as I don't know how to get the stack trace!). It's odd because it was working a few days ago.

After an exhaustive line-by-line binary search I found the problem: when I removed the obfuscated method names I broke a couple of method calls to a generic type. I don't know why it worked with .NET but I guess Mono is a bit more strict about it. Annoyingly, when I did those changes I was using JustDecompile to regularly diff the new patched Shenzhen.exe against the old one to make sure I hadn't broken anything, but it didn't detect this change. The code does show up differently in ILSpy but that doesn't support diffing.

Anyway, I've pushed a fix for it. Did you want to try it out and see if it fixes the crash for you?

**Comment by laurie-bradshaw** [comment]
I initially started doing a similar thing to 12345ieee, but didn't get as far as installing mono dev stuff, as once I'd worked out what the code was doing and realised Shenzhen.exe was the same, I just ran the installer under wine, so I can confirm the patching definitely works on the Linux version.

Presumably the crashing is a build env issue, if gtw123 did a build minus the windowsisms, that'd probably work fine without wine.

**Comment by gtw123** [comment]
@pseuudonym404 It seems like it was a v1.2.0 thing. I think v1.1.0 didn't have the crash.

**Comment by laurie-bradshaw** [comment]
Yeah, saw your post just after I submitted, makes sense. :)

**Comment by gtw123** [comment]
@12345ieee Regarding the other changes:
* I think the only C# 7 features it currently uses are local functions and inline "out" variables. I could remove those if it'll help make it work on other platforms. Would C#6 be OK?
* The backslashes were just me being lazy :)
* Apparently there's a PlatformID.Unix enum - does that work or do you have to use "4"? (I'm guessing it doesn't...)

I'd be happy to merge in those changes if you like, but I don't currently have a way to build binaries for Linux. Is there a way to build Mono binaries on Windows using something like Mono Develop?

**Comment by 12345ieee** [comment]
I confirm that, after patching your latest version to run under linux, everything works.

To see crash reports, go to the save folder, there is a folder called `crash`, that contains the stacktraces.

Should I open a PR for the "de-windowsization" of the installer?

**Comment by gtw123** [comment]
With everyone's help, the installer now supports both Linux and Mac! Released in v1.3.
